### PR TITLE
WIP - Very early first pass at computing distinct counts by aggregation key

### PR DESCRIPTION
**Pull request by amangano-edx**
This is necessary to allow us to filter duplicate course_runs from the response on the front end while still displaying accurate counts.

I had to touch a lot of code to get this to work, but I've tried to keep the changes as isolated as possible to minimize risk. There is still more work to do before this is ready for prod, but for now it would be helpful to get some feedback from the rest of the team on the approach.

@edx/ecommerce

### CODE REVIEW

File: course_discovery/apps/api/v1/views/search.py [hunk 1]
```diff
@@ -118,6 +118,28 @@ class AggregateSearchViewSet(BaseHaystackViewSet):
     facet_serializer_class = serializers.AggregateFacetSearchSerializer
     serializer_class = serializers.AggregateSearchSerializer
 
+    # FIXME Should this be separated and included from an edx-specific app?
+    @list_route(methods=['get'], url_path='distinct_facets')
```
Comment (clintonb) [inline course_discovery/apps/api/v1/views/search.py:122]: None of these method kwargs are necessary.

File: course_discovery/apps/api/v1/views/search.py [hunk 2]
```diff
@@ -118,6 +118,28 @@ class AggregateSearchViewSet(BaseHaystackViewSet):
     facet_serializer_class = serializers.AggregateFacetSearchSerializer
     serializer_class = serializers.AggregateSearchSerializer
 
+    # FIXME Should this be separated and included from an edx-specific app?
```
Comment (clintonb) [inline course_discovery/apps/api/v1/views/search.py:121]: Yeah, this needs to be in the `edx_catalog_extensions` app.

File: course_discovery/apps/api/v1/views/search.py [hunk 3]
```diff
@@ -118,6 +118,28 @@ class AggregateSearchViewSet(BaseHaystackViewSet):
...
+    @list_route(methods=['get'], url_path='distinct_facets')
+    def distinct_facets(self, request):
+        from course_discovery.apps.edx_haystack_extensions.distinct_counts import (DistinctCountsSearchQuerySet,
+                DistinctCountsAggregateFacetSearchSerializer)
+
+        queryset = self.filter_facet_queryset(self.get_queryset())
+        for facet in request.query_params.getlist("selected_facets"):
```
Comment (clintonb) [inline course_discovery/apps/api/v1/views/search.py:128]: Single quotes

File: course_discovery/apps/api/v1/views/search.py [hunk 4]
```diff
@@ -118,6 +118,28 @@ class AggregateSearchViewSet(BaseHaystackViewSet):
...
+            if ":" not in facet:
```
Comment (clintonb) [inline course_discovery/apps/api/v1/views/search.py:129]: What does this mean? Add comments.

File: course_discovery/apps/api/v1/views/search.py (hunk 3 above)
Comment (clintonb) [reply course_discovery/apps/api/v1/views/search.py:128]: Add a comment explaining what's going on here.

File: course_discovery/apps/api/v1/views/search.py [hunk 5]
```diff
@@ -118,6 +118,28 @@ class AggregateSearchViewSet(BaseHaystackViewSet):
     facet_serializer_class = serializers.AggregateFacetSearchSerializer
     serializer_class = serializers.AggregateSearchSerializer
 
+    # FIXME Should this be separated and included from an edx-specific app?
+    @list_route(methods=['get'], url_path='distinct_facets')
+    def distinct_facets(self, request):
+        from course_discovery.apps.edx_haystack_extensions.distinct_counts import (DistinctCountsSearchQuerySet,
```
Comment (clintonb) [inline course_discovery/apps/api/v1/views/search.py:124]: Add a docstring explaining the purpose of this endpoint, and an overview of what it's doing.

File: course_discovery/apps/api/v1/views/search.py [hunk 6]
```diff
@@ -118,6 +118,28 @@ class AggregateSearchViewSet(BaseHaystackViewSet):
...
+                continue
+
+            field, value = facet.split(":", 1)
+            if value:
+                queryset = queryset.narrow('%s:"%s"' % (field, queryset.query.clean(value)))
+
+        queryset = DistinctCountsSearchQuerySet.from_queryset(queryset).set_aggregation_key('aggregation_key_exact')
+
+        serializer_kwargs = {'context': self.get_serializer_context()}
+        serializer_kwargs['context'].update({'objects': queryset, 'many': False})
```
Comment (clintonb) [inline course_discovery/apps/api/v1/views/search.py:139]: This can be  cleaner:

```python
context = self.get_serializer_context()
context.update({'objects': queryset, 'many': False})
serializer = DistinctCountsAggregateFacetSearchSerializer(queryset.facet_counts(), context=context)
```

File: course_discovery/apps/course_metadata/search_indexes.py [hunk 7]
```diff
@@ -110,6 +110,13 @@ class CourseIndex(BaseCourseIndex, indexes.Indexable):
 
     prerequisites = indexes.MultiValueField(faceted=True)
 
+    # A key that can be used to group related documents together to enable the computation of distinct facet and hit
```
Comment (clintonb) [inline course_discovery/apps/course_metadata/search_indexes.py:113]: Why are you touching programs and courses if you're only dealing with course runs?

File: course_discovery/apps/edx_haystack_extensions/distinct_counts.py [hunk 8]
```diff
@@ -0,0 +1,289 @@
+import elasticsearch
```
Comment (clintonb) [inline course_discovery/apps/edx_haystack_extensions/distinct_counts.py:1]: This needs to be broken up appropriately:

* distinct_counts/backends.py
* distinct_counts/serializers.py
* distinct_counts/query.py

**Review by clintonb — COMMENTED** [review]
I'm honestly not sure what's going on here. Before adding more code, we need to document the expected behavior of this new endpoint. What are the inputs? What's the expected output? Once you have that document what's going on with aggregations.

tl;dr Less code, more English.

**Comment by clintonb** [comment]
P.S. Add the ticket number to your commit message.

File: course_discovery/apps/api/v1/views/search.py (hunk 3 above)
Comment (MatthewPiatetsky) [inline course_discovery/apps/api/v1/views/search.py:128]: can you add a comment explaining this?

File: course_discovery/apps/api/v1/views/search.py (hunk 5 above)
Comment (MatthewPiatetsky) [inline course_discovery/apps/api/v1/views/search.py:124]: i would put the imports at the top

File: course_discovery/apps/course_metadata/search_indexes.py (hunk 7 above)
Comment (amangano-edx) [reply course_discovery/apps/course_metadata/search_indexes.py:113]: because the marketing site expects to get back both courseruns and programs with the response, and in order for ES to compute the distinct counts correctly, all records need to have a value for the aggregation key.

File: course_discovery/apps/api/v1/views/search.py (hunk 1 above)
Comment (amangano-edx) [reply course_discovery/apps/api/v1/views/search.py:122]: This is doing the same thing as the `facets` method in this file and in the drf_haystack mixins module.

File: course_discovery/apps/api/v1/views/search.py (hunk 2 above)
Comment (MatthewPiatetsky) [reply course_discovery/apps/api/v1/views/search.py:121]: why is getting facet results based on deduped data edx specific? or would you say that because you think search for multiple course runs isn't something we'll support now for open source users

File: course_discovery/apps/api/v1/views/search.py (hunk 3 above)
Comment (amangano-edx) [reply course_discovery/apps/api/v1/views/search.py:128]: This follows what is being done in the drf_haystack `facets` endpoint method.

File: course_discovery/apps/api/v1/views/search.py (hunk 4 above)
Comment (amangano-edx) [reply course_discovery/apps/api/v1/views/search.py:129]: This follows what is being done in the drf_haystack facets endpoint method.

File: course_discovery/apps/api/v1/views/search.py (hunk 3 above)
Comment (amangano-edx) [reply course_discovery/apps/api/v1/views/search.py:128]: This follows what is being done in the drf_haystack facets endpoint method.

File: course_discovery/apps/api/v1/views/search.py (hunk 5 above)
Comment (amangano-edx) [reply course_discovery/apps/api/v1/views/search.py:124]: I just put it here to keep all the hacky code together. I will move the imports to the top when I move this method into the edx_catalog_extensions app.

File: course_discovery/apps/api/v1/views/search.py (hunk 2 above)
Comment (amangano-edx) [reply course_discovery/apps/api/v1/views/search.py:121]: It's an edx-specific hack that we do not want to be exposed to the general community, so I agree with Clinton that it should be in an edx-specific app.

**Comment by amangano-edx** [comment]
replaced by https://github.com/edx/course-discovery/pull/704
